File: codeforces/1800~1899/1896/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0, q = 0;
  cin >> n >> q;
  vector<int> a(n);
  for (int i = 0; i < n; ++i) cin >> a[i];

  set<int> one;
  for (int i = 0; i < n; ++i) {
    if (a[i] == 1) {
      one.insert(i);
    }
  }

  for (int k = 0; k < q; ++k) {
    int type = 0;
    cin >> type;
    if (type == 1) {
      int s = 0;
      cin >> s;
      if (one.empty()) {
        int sum = n * 2;
        if (s > sum || s % 2 != 0) {
          cout << "NO" << endl;
        } else {
          cout << "YES" << endl;
        }
      } else {
        int one_count = one.size();
        if (s % 2 != one_count % 2) {
          int s1 = 2 * (*one.rbegin()) - one_count + 1;
          int s2 = 2 * (n - *one.begin() - 1) - one_count + 1;
          if (s > max(s1, s2)) {
            cout << "NO" << endl;
          } else {
            cout << "YES" << endl;
          }
        } else {
          if (s > 2 * n - one_count) {
            cout << "NO" << endl;
          } else {
            cout << "YES" << endl;
          }
        }
      }
    } else if (type == 2) {
      int i = 0, v = 0;
      cin >> i >> v;
      i--;
      one.erase(i);
      a[i] = v;
      if (a[i] == 1) {
        one.insert(i);
      }
    }
  }
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1800~1899/1896/d.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::BTreeSet;
use std::cmp::max;
fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());

    let inputs = read_vec::<usize>();
    let n = inputs[0];
    let q = inputs[1];
    let mut a = read_vec::<i32>();
    let mut one = BTreeSet::new();
    for i in 0..n {
        if a[i] == 1 {
            one.insert(i);
        }
    }

    for _ in 0..q {
        let query = read_vec::<usize>();
        if query[0] == 1 {
            let s = query[1];
            if one.is_empty() {
                let sum = n * 2;
                if s > sum || s % 2 != 0 {
                    writeln!(writer, "NO").ok();
                } else {
                    writeln!(writer, "YES").ok();
                }
            } else {
                let one_count = one.len();
                if one_count % 2 != s % 2 {
                    let s1 = 2 * (*one.last().unwrap()) - one_count + 1;
                    let s2 = 2 * (n - *one.first().unwrap() - 1) - one_count + 1;
                    if s > max(s1, s2) {
                        writeln!(writer, "NO").ok();
                    } else {
                        writeln!(writer, "YES").ok();
                    }
                } else {
                    let sum = n * 2 - one_count;
                    if s > sum {
                        writeln!(writer, "NO").ok();
                    } else {
                        writeln!(writer, "YES").ok();
                    }
                }
            }
        } else if query[0] == 2 {
            let i = query[1]-1;
            one.remove(&i);
            a[i] = query[2] as i32;
            if a[i] == 1 {
                one.insert(i);
            }
        }
    }
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
